lib/compress/zstd_lazy.c

```diff
@@ -845,11 +860,633 @@ FORCE_INLINE_TEMPLATE size_t ZSTD_HcFindBestMatch_extDict_selectMLS (
+        U32* const row = hashTable + relRow;
+        BYTE* tagRow = (BYTE*)(tagTable + relRow);
+        U32 const head = *tagRow & rowMask;
+        U32 matchBuffer[32 /* maximum nb row entries */];
+        size_t numMatches = 0;
+        size_t currMatch = 0;
+        ZSTD_VecMask matches = ZSTD_row_getMatchMask(tagRow, (BYTE)tag, head, rowEntries);
+
+        /* Cycle through the matches and prefetch */
+        for (; (matches > 0) && (nbAttempts > 0); --nbAttempts, matches &= (matches - 1)) {
+            U32 const matchPos = (head + ZSTD_VecMask_next(matches)) & rowMask;
+            U32 const matchIndex = row[matchPos];
+            if (matchIndex < lowLimit)
+                break;
+            if ((dictMode != ZSTD_extDict) || matchIndex >= dictLimit) {
+                PREFETCH_L1(base + matchIndex);
+            } else {
+                PREFETCH_L1(dictBase + matchIndex);
+            }
+            matchBuffer[numMatches++] = matchIndex;
+        }
+        
+        /* Speed opt: insert current byte into hashtable too. This allows us to avoid one iteration of the loop
+           in ZSTD_row_update_internal() at the next search. */
+        {
+            U32 const pos = ZSTD_row_nextIndex(tagRow, rowMask);
+            tagRow[pos + ZSTD_ROW_HASH_TAG_OFFSET] = (BYTE)tag;
+            row[pos] = ms->nextToUpdate++;
+        }
+
+        /* Return the longest match */
+        for (; currMatch < numMatches; ++currMatch) {
+            U32 const matchIndex = matchBuffer[currMatch];
+            size_t currentMl=0;
+            assert(matchIndex < curr);
+            assert(matchIndex >= lowLimit);
+
+            if ((dictMode != ZSTD_extDict) || matchIndex >= dictLimit) {
+                const BYTE* const match = base + matchIndex;
+                assert(matchIndex >= dictLimit);   /* ensures this is true if dictMode != ZSTD_extDict */
+                if (match[ml] == ip[ml])   /* potentially better */
+                    currentMl = ZSTD_count(ip, match, iLimit);
+            } else {
+                const BYTE* const match = dictBase + matchIndex;
+                assert(match+4 <= dictEnd);
+                if (MEM_read32(match) == MEM_read32(ip))   /* assumption : matchIndex <= dictLimit-4 (by table construction) */
+                    currentMl = ZSTD_count_2segments(ip+4, match+4, iLimit, dictEnd, prefixStart) + 4;
+            }
+
+            /* Save best solution */
+            if (currentMl > ml) {
+                ml = currentMl;
+                *offsetPtr = curr - matchIndex + ZSTD_REP_MOVE;
+                if (ip+currentMl == iLimit) break; /* best possible, avoids read overflow on next attempt */
+            }
+        }
+    }
+
+    if (dictMode == ZSTD_dedicatedDictSearch) {
+        ml = ZSTD_dedicatedDictSearch_lazy_search(offsetPtr, ml, nbAttempts, dms,
+                                                  ip, iLimit, prefixStart, curr, dictLimit, ddsIdx);
+    } else if (dictMode == ZSTD_dictMatchState) {
```

Comment:
@senhuang42 since all the tests were passing with a buggy DMS implementation, can you add a test case that would have caught it? Or show that the LLVM/zstd fuzzers would catch it? And then update the summary.

Reply:
Also, when does the DMS get used? If it is used in scenarios we care about, we should investigate adding prefetching. I would be okay getting a correct implementation in this PR, and adding prefetching later. But, if that is the case, can you add a TODO to the code (and the PR summary) so we can remember?

Reply:
Yeah, the DMS was failing silently cause everything in the hashtable would technically still be valid indices, just not in any recognizable format to non-rowhash matchfinders. 

So the DMS codepath happens when: `ZSTD_matchState_dictMode()` gives you DMS and `ZSTD_shouldAttachDict()` are true. 

Something like:
```
ZSTD_CCtx_loadDictionary(ctxOrig, dict, flatdictSize);
ZSTD_compress2(ctxOrig, compressedBuffer, compressedBufferSize, contentStart, contentSize);
```
seems to do it.
I'll expand the `dictionary efficiency test` in `fuzzer.c` to include this way of compressing using a dict.

Just based on how often DMS gets triggered in fuzzer.c, doesn't seem to be that common? (and isn't DDS a superior successor to it?) For some examples: the following unit tests use dms:
`test112 : compress with static CDict`
`test126 : Loading a dictionary clears the prefix`
`test127 : Loading a dictionary clears the cdict`

I didn't include prefetching at all since it seemed like it was not good on smaller files/smaller hashtables, and the DMS typically aren't huge - but in either case, yeah, we should measure that.

Reply:
> I didn't include prefetching at all since it seemed like it was not good on smaller files/smaller hashtables, and the DMS typically aren't huge - but in either case, yeah, we should measure that, and I'll add it todo.

Prefetching in the DMS is very different then prefetching for a the file's hash table. We're inserting into the hash table as we compress, so those writes mean it is hot and likely already in L1.

However, the dictionary is read-only, and likely starts out in RAM (cold). So we'll see lots of cache misses. @felixhandte found that prefetching was very? beneficial for DDS.

Reply:
The speed gain of adding a prefetch of the hashtable row and tagRow definitely helps DMS a bit, I've seen around 5-10% faster. Prefetching the match appears to provide a tiny bump to speed.